Zend/zend_API.c

```diff
@@ -2032,6 +2032,59 @@ static void zend_check_magic_method_args(
 	}
 }
 
+static const char *zend_get_return_type_by_const(int type)
+{
+	switch(type) {
+		case MAY_BE_STRING:
+			return "string";
+		case MAY_BE_BOOL:
+			return "bool";
+		case MAY_BE_OBJECT:
+			return "object";
+		case MAY_BE_ARRAY:
+			return "array";
+		case (MAY_BE_ARRAY | MAY_BE_NULL):
+			return "?array";
+		case MAY_BE_VOID:
+			return "void";
+		EMPTY_SWITCH_DEFAULT_CASE()
+	}
+}
+
+static void zend_check_magic_method_arg_type(uint32_t arg_num, const zend_class_entry *ce, const zend_function *fptr, int error_type, int arg_type)
+{
+		if (
+			ZEND_TYPE_IS_SET(fptr->common.arg_info[arg_num].type)
+			 && ! (ZEND_TYPE_FULL_MASK(fptr->common.arg_info[arg_num].type) & arg_type)
+		) {
+			zend_error(error_type, "%s::%s(): Argument #%d ($%s) must be of type %s when declared",
+				ZSTR_VAL(ce->name), ZSTR_VAL(fptr->common.function_name),
+				arg_num, ZSTR_VAL(fptr->common.arg_info[arg_num].name), zend_get_return_type_by_const(arg_type));
+		}
+}
+
+static void zend_check_magic_method_return_type(const zend_class_entry *ce, const zend_function *fptr, int error_type, int return_type)
+{
+	if (
+		(fptr->common.fn_flags & ZEND_ACC_HAS_RETURN_TYPE)
+		&& (ZEND_TYPE_FULL_MASK(fptr->common.arg_info[-1].type) != return_type)
```

Comment:
Couldn't `zend_check_magic_method_return_type()` and `zend_check_magic_method_union_return_type()` be consolidated into a single function?

Reply:
There are a little bit different in their checks:

```c
// single type
&& (ZEND_TYPE_FULL_MASK(fptr->common.arg_info[-1].type) != return_type)
```

vs.

```c
// union type
&& ! (ZEND_TYPE_FULL_MASK(fptr->common.arg_info[-1].type) & union_return_type)
```

But I'll try to combine them 👍

Reply:
Yeah, I saw the difference, but it should be possible to combine them

Reply:
I think it's okay to keep them separate, because they have different variance, so it might be more confusing to combine them.

Reply:
Oh sorry, I misread that this is about the two return type functions. Yeah, those should be combined.

Reply:
> Yeah, I saw the difference, but it should be possible to combine them

Any idea how? Looks like if we use `!= return_type`, it breaks the union case types and if we use `& union_return_type` it allows standalone signatures like

```php
<?php
class Foo {
	function __isset($name) : \stdClass|bool {}
}
```

which is not desired 😕

Reply:
What you want is "has any type outside the allowed ones". The way to do that would be something like `if (ZEND_TYPE_FULL_MASK() & ~return_type) { error }`

Reply:
Thank you both, implementation is even cleaner now 🎉